community/primitive-collections/src/test/java/org/neo4j/collection/primitive/PrimitiveLongResourceCollectionsTest.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.collection.primitive;
+
+import org.junit.Test;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.LongPredicate;
+
+import org.neo4j.graphdb.Resource;
+
+import static org.junit.Assert.assertEquals;
+
+public class PrimitiveLongResourceCollectionsTest
```

Comment:
I get the impression that the iterators do not close their resources when exhausted. Is this the expected behavior?

Reply:
You are correct, they don't close when exhausted, nice catch! I've actually not thought about that for the iterators from `PrimitiveLongResourceCollections`.
The `PrimitiveLongResourceIterator` interface does not say anything about such behaviour explicitly, but what other behaviour would you realistically want when you explicitly create a resource iterator?

Pro
 - Good safety net for when you forget to close
 - Convenient to not be forced to close if you exhaust it (which you most often do)

Con
 - Allow faulty code and hide bugs.
 - More side-effects -> less control and harder to reason about.

For the general purpose of `PrimitiveLongResourceCollections` I think the cons are bigger than the pros.

Side note: Label and schema index iterators do close their resources on exhaust for legacy reasons. Up until now this was the only way to make sure they where closed at all.